pkg/sql/ordering.go

```diff
@@ -24,35 +24,83 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/encoding"
 )
 
-// orderingInfo describes the column ordering on a set of results.
+// orderingInfo describes the column ordering on a set of results. Typically the
+// ordering information is used to see to what extent it can satisfy a "desired"
+// ordering (which is just a list of columns and directions).
 //
-// If results are known to be restricted to a single value on some columns, we call these "exact
-// match" columns; these are inconsequential w.r.t ordering.
+// In its simplest form, an ordering is simply a list of columns and a direction
+// for each column, for example a+,b-,c+. This indicates that the rows are
+// ordered by the value on column a (ascending); rows that have the same value
+// on a are ordered by column b (descending); and rows that have the same values
+// on a and b are ordered by column c (ascending).
 //
-// For example, if an index was defined on columns (a, b, c, d) and the WHERE clause was
-// "(a, c) = (1, 2)" then a and c are exact-match columns and we have an ordering by b then d.
-// Such an ordering satisfies any of the following desired orderings (among many others):
-//    a, c
-//    c, a
-//    b, a, c
-//    b, c, a
-//    a, b, c
-//    c, b, a
-//    b, d, a
-//    a, b, c, d
-//    b, a, c, d
+// == Exact match columns ==
 //
+// If results are known to be restricted to a single value on some columns, we
+// call these "exact match" columns; these are inconsequential w.r.t ordering.
+//
+// For example, if an index was defined on columns (a, b, c, d) and the WHERE
+// clause was "(a, c) = (1, 2)" then a and c are exact-match columns and we have
+// an ordering by b+ then d+. Such an ordering satisfies any of the following
+// desired orderings (among many others):
+//   a+,c+
+//   a-,c+
+//   b+,a+,c+
+//   b+,c+,a-
+//   a+,b+,c+
+//   c-,b+,a+
+//   b+,d+,a-
+//   a+,b+,c+,d+
+//   b+,a+,c-,d+
+//
+// == Unique flag ==
+//
+// In some cases (like unique indexes), we have an ordering on a set of columns
+// and we know that each row has a unique set of values on these columns. For
+// example, if we have a unique index on columns (a, b), the results have the
+// ordering a+,b+ and the unique flag tells us that this satisfies any desired
+// ordering that has a+,b+ as a prefix (such as a+,b+,c-,d+).
```

Comment:
Yes I think we should have that kind of infrastructure. It will be even more useful when we implement a better structure which allows multiple "keys" (like the one described in the paper above).